pycytominer/cyto_utils/cells.py

```diff
@@ -430,46 +435,210 @@ def aggregate_compartment(
-                    columns={f"{self.fields_of_view_feature}": f"Metadata_Site_Count"}
+            if compute_counts and self.fields_of_view_feature not in self.strata:
+                fields_count_df = self.image_df.loc[
+                    :, list(np.union1d(self.strata, self.fields_of_view_feature))
+                ]
+                fields_count_df = (
+                    fields_count_df.groupby(self.strata)[self.fields_of_view_feature]
+                    .count()
+                    .reset_index()
+                    .rename(
+                        columns={
+                            f"{self.fields_of_view_feature}": f"Metadata_Site_Count"
+                        }
+                    )
                 )
-            )
 
-            object_df = fields_count_df.merge(object_df, on=self.strata, how="right")
+                partial_object_df = fields_count_df.merge(
+                    partial_object_df,
+                    on=self.strata,
+                    how="right",
+                )
+
+            object_dfs.append(partial_object_df)
+
+        # Concatenate one or more aggregated dataframes row-wise into final output
+        object_df = pd.concat(object_dfs, axis=0).reset_index(drop=True)
 
         return object_df
 
+    def _compartment_df_generator(
+        self,
+        compartment,
+        n_strata=1,
+    ):
+        """A generator function that returns chunks of the entire compartment
+        table from disk.
+
+        We want to return dataframes with all compartment entries within unique
+        combinations of self.merge_cols when aggregated by self.strata
+
+        Parameters
+        ----------
+        compartment : str
+            Compartment to aggregate.
+        n_strata : int, default 1
+            Number of unique strata to pull from the database into working memory
+            at once.  A larger number does not lead to faster compute times
+
+        Yields
+        ------
+        image_df : pandas.core.frame.DataFrame
+            A generator whose __next__() call returns a chunk of the compartment
+            table, where rows comprising a unique aggregation stratum are not split
+            between chunks, and thus groupby aggregations are valid
+
+        """
+
+        assert (
+            n_strata > 0
+        ), "Number of strata to pull into memory at once (n_strata) must be > 0"
+
+        # Columns of the compartment table we will obtain
+        cols = "*"
+        compartment_row1 = pd.read_sql(
+            sql=f"select {cols} from {compartment} limit 1",
+            con=self.conn,
+        )
+        typeof_str = ", ".join([f"typeof({x})" for x in compartment_row1.columns])
+        compartment_dtypes = pd.read_sql(
+            sql=f"select {typeof_str} from {compartment} limit 1",
+            con=self.conn,
+        )
+        dtype_dict = dict(
+            zip(
+                [s[7:-1] for s in compartment_dtypes.columns],  # strip typeof( )
```

Comment:
hmm, i don't know what is going on here. You're pulling in only columns of a certain dtype, finding their names, and then creating a dictionary such that you do not need to infer dtypes in subsequent `pandas.read_sql()` calls?

Why do you need to subset [7:-1]? I am always weary of these hardcoded calls!

Reply:
This is doing the following:
1. get the first row of the `compartment` table from the SQLite, so that I know all the column names
2. get the `dtype` of every column in the SQLite (which I have to request by name... that's why I needed the names)

These `dtype`s are needed to construct valid queries (specifically, `TEXT` must include quotations).

The subset `[7:-1]` I agree is ugly... the reason for it is that the columns come back with names like `typeof(feature1)` instead of `feature1`.  I wanted to strip out "typeof(" from the beginning and ")" from the end.  The columns always come back from the SQL request named in exactly this way... so I think the code is robust... but it is hard to understand.

Reply:
It might be possible to replace
```python
            zip(
                [s[7:-1] for s in compartment_dtypes.columns],  # strip typeof( )
                compartment_dtypes.iloc[0].values,
            )
```
with
```python
            zip(
                [s for s in compartment_row1.columns],
                compartment_dtypes.iloc[0].values,
            )
```
but I am not sure that the columns _always_ come back in the same order as the request.

Reply:
I don't know of a way to make a request that doesn't return the information in this way.  One thing I could do to improve readability is to make a little helper function or lambda function whose purpose is clearly and explicitly to remove `typeof()`... that way it's more readable.  For example, replace this

```python
            zip(
                [s[7:-1] for s in compartment_dtypes.columns],  # strip typeof( )
                compartment_dtypes.iloc[0].values,
            )
```
with this
```python
            # strip the characters "typeof(" from the beginning and ")" from the end of 
            # column names returned by SQLite
            strip_typeof = lambda s: s[7:-1]  
            zip(
                [strip_typeof(s) for s in compartment_dtypes.columns],  
                compartment_dtypes.iloc[0].values,
            )
```